components/brave_ads/core/internal/resources/resources_util_impl.h

```diff
@@ -27,32 +27,33 @@ class File;
 namespace brave_ads::resource {
 
 template <typename T>
-std::unique_ptr<ParsingResult<T>> ReadFileAndParseResourceOnBackgroundThread(
+base::expected<T, std::string> ReadFileAndParseResourceOnBackgroundThread(
     base::File file) {
   if (!file.IsValid()) {
-    return {};
-  }
-  std::string content;
-  const base::ScopedFILE stream(base::FileToFILE(std::move(file), "rb"));
-  if (!base::ReadStreamToString(stream.get(), &content)) {
-    return {};
+    return base::unexpected("File is not valid.");
   }
 
-  absl::optional<base::Value> root = base::JSONReader::Read(content);
-  if (!root) {
-    return {};
-  }
+  absl::optional<base::Value> root;
+  {
+    std::string content;
+    const base::ScopedFILE stream(base::FileToFILE(std::move(file), "rb"));
+    if (!base::ReadStreamToString(stream.get(), &content)) {
+      return base::unexpected("Couldn't read file.");
+    }
 
-  // Free memory in advance to optimize the peak memory consumption. |json| can
-  // be up to 10Mb and the following code allocates an extra few Mb of memory.
-  content = std::string();
+    root = base::JSONReader::Read(content);
+    if (!root) {
+      return base::unexpected("Failed to parse json.");
+    }
+    // `content` can be up to 10MB, so we keep, so scope it to this block, to
+    // release its memory as soon as possible.
```

Comment:
When I first read I was confused by `so we keep`, maybe `|content| can be up to 10 MB, so we keep the scope of it to this block to release its memory as soon as possible.`

Reply:
Done.